<a href="https://www.kaggle.com/code/sahidvelji/cleaning-the-ontario-sunshine-list-data?scriptVersionId=100028357" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

# Cleaning The Ontario Sunshine List Data

<a id="table-of-contents"></a>

This notebook was used to clean Ontario's public sector salary disclosure data (also known as the Ontario sunshine list). The cleaned dataset can be found [here](https://www.kaggle.com/sahidvelji/the-ontario-sunshine-list), and an EDA of the 2019 data can be found [here](https://www.kaggle.com/sahidvelji/the-ontario-sunshine-list-2019-eda).

## Table of contents

<p style="line-height: 1.6em;">
    <a href="#loading-data">1. Loading the data</a><br>
    <a href="#renaming-columns">2. Renaming columns</a><br>
    <a href="#data-types">3. Data types</a><br>
    <a href="#missing-values">4. Missing values</a><br>
    <a href="#concat-data">5. Concatenating the dataframes</a><br>
    <a href="#write-to-csv">6. Writing to CSV</a><br>
</p>

In [1]:
import os
import re
import pandas as pd
import numpy as np
import shutil
from IPython.display import display, HTML
import plotly.express as px
from pathlib import Path

YEARS = range(1996, 2020)

px.defaults.template = 'plotly_white'
px.defaults.color_discrete_sequence = ['steelblue']
MODE_BAR_BUTTONS = ['toImage', 'zoom2d', 'pan2d', 'select2d', 'lasso2d',
                    'zoomIn2d', 'zoomOut2d', 'autoScale2d', 'resetScale2d',
                    'toggleSpikelines', 'hoverClosestCartesian', 'hoverCompareCartesian']
CONFIG = {
    'modeBarButtonsToRemove': ['pan2d', 'select2d', 'lasso2d', 'toggleSpikelines']
}

# Loading the data

In [2]:
path = Path('/kaggle/input/the-ontario-sunshine-list-raw-data')  
filenames = os.listdir(path)
filenames

['en-2018-pssd-compendium.csv',
 'en-2003-pssd.csv',
 'en-2004-pssd.csv',
 'en-2016-pssd-compendium-20171128-utf8.csv',
 'en-2006-pssd.csv',
 'en-1999-pssd.csv',
 'en-2018-pssd-compendium-20191223.csv',
 'en-2002-pssd.csv',
 'en-2008-pssd.csv',
 'pssd-en-2013.csv',
 'en-2000-pssd.csv',
 'en-2007-pssd.csv',
 'en-2009-pssd.csv',
 'en-1997-pssd.csv',
 '2010.csv',
 'en-2005-pssd.csv',
 'tbs-pssd-compendium-en-utf8-2019.csv',
 'en-1998-pssd.csv',
 '2014-pssd-full-compendium-utf8-en.csv',
 'en-2001-pssd.csv',
 '2011_0.csv',
 'en-1996-pssd.csv',
 'pssd-en-2012-fixed_0.csv',
 'en-2015-pssd-compendium-with-addendum-20161219.csv']

While downloading the data from the government of Ontario website, I realized that the filenames were inconsistent. Even worse, there are two files that appear to both be data from 2018: `en-2018-pss-compendium.csv` and `en-2018-pss-compendium-20191223.csv`. Let's examine these two files.

In [3]:
pd.read_csv(path/'en-2018-pssd-compendium.csv', nrows=5)

Sector Last Name    First Name  Salary Paid Taxable Benefits  \
0  Colleges  Abiscott         Alexa  $197,073.24        $2,033.69   
1  Colleges       Ahn       Song Ho  $114,331.63           $51.57   
2  Colleges    Aitken        Sharon  $118,650.76        $1,250.00   
3  Colleges       Ali       Shirook  $104,398.32          $101.50   
4  Colleges   Allcott  Austin Micha  $124,307.70        $1,264.54   

                                            Employer  \
0  Sheridan College Institute of Technology and A...   
1  Sheridan College Institute of Technology and A...   
2  Sheridan College Institute of Technology and A...   
3  Sheridan College Institute of Technology and A...   
4  Sheridan College Institute of Technology and A...   

                                         Job Title  Calendar Year  
0  General Counsel And Information Privacy Officer           2017  
1     Visualization Researcher - Part-Time Faculty           2017  
2                Director Development and Campaign           2017  
3                                        Professor           2017  
4                                             Dean           2017

In [4]:
pd.read_csv(path/'en-2018-pssd-compendium-20191223.csv', nrows=5)

Sector    Last Name First Name   Salary Paid Taxable Benefits  \
0  Colleges        Jensen     Cheryl  $ 337,533.24       $ 3,408.12   
1  Colleges         Brule     Claude  $ 230,976.76         $ 374.96   
2  Colleges   Wotherspoon       Doug  $ 230,788.38         $ 186.58   
3  Colleges        Mcnair      Duane  $ 221,744.64         $ 366.12   
4  Colleges       Stanbra      Laura  $ 221,744.64         $ 366.12   

            Employer                                  Job Title  Calendar Year  
0  Algonquin College                                  President           2018  
1  Algonquin College             Senior Vice President Academic           2018  
2  Algonquin College     Vice President Innovation and Strategy           2018  
3  Algonquin College  Vice President Finance and Administration           2018  
4  Algonquin College            Vice President Student Services           2018

It turns out that `en-2018-pss-compendium.csv` is actually data from 2017 and `en-2018-pss-compendium-20191223.csv` is data from 2018. The input directory is read-only data, so we cannot rename files here. However, we will use a dictionary to organize the files by year after loading them into dataframes. First, we'll create a regular expression to extract the year from each filename. 

In [5]:
p = re.compile('(?:^|-)(\d{4})\D')

That should do it. The year appears either at the beginning of the filename or after a dash. The character following the year is either a dash, a dot, or an underscore. This means we have a non-numeric character following the year. The regular expression was constructed based on these observations: first we match the beginning of a string or a dash in a non-capturing group. Then, we match 4 digits in a capturing group. This is followed by a single non-numeric digit.

Unfortunately, we encounter a `UnicodeDecodeError` if we try to load the data with the standard utf-8 encoding for some of the files. Instead, we'll use latin1 encoding if the default fails. The error occurs due to characters such as "é" in the data.

In [6]:
def read_csv(filename):
    try:
        return pd.read_csv(path/filename, encoding='utf-8')
    except UnicodeDecodeError:
        return pd.read_csv(path/filename, encoding='latin1')

In [7]:
pss = {}
for filename in filenames:
    if filename == 'en-2018-pssd-compendium.csv':
        pss[2017] = read_csv(filename)
        continue
    m = p.search(filename)
    year = int(m.group(1))
    pss[year] = read_csv(filename)

Now we have a dictionary where each key is a calendar year and the value is the corresponding dataframe. The `describe` method is a useful way of displaying a summary table of the data. Examining all 24 tables one by one wouldn't be very efficient but we'll display the tables here for reference purposes anyways.

In [8]:
for year in YEARS:
    display(HTML(pss[year].describe(include='all').to_html()))

The Salary Paid and Taxable Benefits columns should have type `float` but it looks like they are currently strings because of the dollar signs and commas. We will fix this later.

First, we will make sure that the column names of all dataframes match.

Second, we will deal with the data types of the columns. As mentioned above, we expect the Salary Paid and Taxable Benefits columns to have type `float`. The Calendar Year column should have type `int` and the rest of the columns should have type `string`.

Then, we will check for missing values and see if there is a good way of imputing them. Finally, we will explore the possibility of concatenating the dataframes. After all, they should have the same number of columns, the same column names and data types, and one dataframe is easier to work with than 24 separate ones.

<a id="renaming-columns"></a>
[Return to table of contents](#table-of-contents)

# Renaming columns

We will check to make sure that every dataframe has the same column names using the 2019 columns as a reference.

In [9]:
refcols = pss[2019].columns
for year in YEARS:
    if not refcols.equals(pss[year].columns):
        print(year, pss[year].columns.tolist(), sep='\n', end='\n\n')

print("2019", refcols.tolist(), sep='\n')

1996
['Sector', 'Last Name', 'First Name', 'Salary Paid', 'Taxable Benefits', 'Employer', 'Job Title', 'Calendar Year', 'Unnamed: 8']

2001
['Sector', 'Surname', 'First Name', 'Salary Paid', 'Taxable Benefits', 'Employer', 'Position', 'Calendar Year']

2009
['Sector', 'Last Name', 'First Name', 'Salary Paid ', 'Taxable Benefits', 'Employer', 'Job Title', 'Calendar Year']

2010
['Sector', 'Last Name', 'First Name', 'Salary Paid ', 'Taxable Benefits', 'Employer', 'Job Title', 'Calendar Year']

2011
['Sector', 'Last Name', 'First Name', 'Salary Paid ', 'Taxable Benefits', 'Employer', 'Job Title', 'Calendar Year']

2014
['Sector', 'Last name', 'First Name', 'Salary Paid', 'Taxable Benefits', 'Employer', 'Job title', 'Calendar year']

2019
['Sector', 'Last Name', 'First Name', 'Salary Paid', 'Taxable Benefits', 'Employer', 'Job Title', 'Calendar Year']


There are numerous inconsistencies:

- There is an extra column in the 1996 dataframe. 
- The 2001 dataframe has Surname instead of Last Name and Position instead of Job Title. 
- There are trailing whitespaces in the Salary Paid column in the 2009, 2010, and 2011 dataframes.
- In the 2014 dataframe, the second words of "Last name", "Job title", and "Calendar year" are not capitalized. 

We will first examine the extra column in the 1996 dataframe and drop it if appropriate.

In [10]:
pss[1996].head()

Sector Last Name First Name  Salary Paid  \
0  OTHER PUBLIC SECTOR EMPLOYERS   Kendall      Perry  $194,890.40   
1  OTHER PUBLIC SECTOR EMPLOYERS      Rehm    Juergen  $115,603.62   
2  OTHER PUBLIC SECTOR EMPLOYERS      Room      Robin  $149,434.48   
3         ONTARIO PUBLIC SERVICE      KNOX      KEN W  $109,382.92   
4                      HOSPITALS     Cliff      Bruce  $110,309.00   

  Taxable Benefits                             Employer  \
0          $711.24        Addiction Research Foundation   
1          $403.41        Addiction Research Foundation   
2          $512.58        Addiction Research Foundation   
3        $4,921.68   Agriculture,Food and Rural Affairs   
4        $3,157.00  Ajax and Pickering General Hospital   

                                           Job Title  Calendar Year  \
0                                    President & CEO           1996   
1  Dir., Soc. Eval. Research & Act. Dir., Clin. R...           1996   
2     V.P., Research & Coordinator, Intern. Programs           1996   
3                                    Deputy Minister           1996   
4                                    President & CEO           1996   

   Unnamed: 8  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN

In [11]:
pss[1996]['Unnamed: 8'].isna().all()

True

This unnamed column is filled with `NaN` values, meaning that we can safely drop this column.

In [12]:
pss[1996] = pss[1996].drop(columns='Unnamed: 8')

Now that every dataframe has the same number of columns, let's rename all columns to match the columns of the 2019 dataframe.

In [13]:
for year in YEARS:
    pss[year].columns = refcols

The Calendar Year column may seem redundant because of the file name. For example, we know that all of the data in `tbs-pssd-compendium-en-utf8-2019.csv` is for the calendar year 2019. However, we will not drop this column yet in case we later want to concatenate the dataframes.

<a id="data-types"></a>
[Return to table of contents](#table-of-contents)

# Data types

Here, we will examine the data types of the columns and ensure that they match across all dataframes.

In [14]:
for year in YEARS:
    print(year, pss[year].dtypes, sep='\n', end='\n\n')

1996
Sector              object
Last Name           object
First Name          object
Salary Paid         object
Taxable Benefits    object
Employer            object
Job Title           object
Calendar Year        int64
dtype: object

1997
Sector              object
Last Name           object
First Name          object
Salary Paid         object
Taxable Benefits    object
Employer            object
Job Title           object
Calendar Year        int64
dtype: object

1998
Sector              object
Last Name           object
First Name          object
Salary Paid         object
Taxable Benefits    object
Employer            object
Job Title           object
Calendar Year        int64
dtype: object

1999
Sector              object
Last Name           object
First Name          object
Salary Paid         object
Taxable Benefits    object
Employer            object
Job Title           object
Calendar Year        int64
dtype: object

2000
Sector              object
Last Name           obje

Unfortunately, we discover more inconsistencies. Quickly scrolling through the output tells us that for most dataframes, the Calendar Year column has data type `int`, except for the 2016 dataframe. Also, the dataframes for 2012 and 2013 have type `float` for the Salary Paid and Taxable Benefits columns while the other dataframes have type `object` for these columns.

Let's find out why the Calendar Year column in the 2016 dataframe doesn't have type `int`.

In [15]:
pss[2016]['Calendar Year'].nunique()

38

That doesn't look promising. We expect a single unique value for the Calendar Year column, but we have 38 unique values instead.

In [16]:
wrong_year = pss[2016][pss[2016]['Calendar Year'] != '2016']
wrong_year.shape[0]

65

In [17]:
wrong_year.head(20)

Sector       Last Name     First Name  \
6223                           Crown Agencies           Fagan         Thomas   
9323                           Crown Agencies           Doran        Patrick   
23040      Government of Ontario - Ministries         LeBlanc         Laurie   
32225   Hospitals and Boards of Public Health           Levac        Jody J.   
32238   Hospitals and Boards of Public Health        Mccarron    Veronica J.   
32319   Hospitals and Boards of Public Health           Wells  Antoinette L.   
110271                           Universities        Phillips       Paulette   
115604                           Universities          Bender    Daniel Eric   
116990                           Universities        Kingston           Paul   
117503                           Universities           Menou        Kristen   
118209                           Universities           Segal         Zindel   
121619                           Universities            Adam          Barry   
121638                           Universities           Arbex        Marcelo   
121647                           Universities           Bacon         Tamsin   
121660                           Universities           Basok          Tanya   
121666                           Universities        Berryman        Jeffrey   
121686                           Universities         Calarco         Pascal   
121694                           Universities  Cavallo-Medved           Dora   
121706                           Universities          Cheung      Chi-Keung   
121727                           Universities          Dalton           Joan   

        Salary Paid Taxable Benefits  \
6223    $126,498.77            $0.00   
9323    $121,415.63            $0.00   
23040   $242,431.56       $23,569.27   
32225   $169,367.20          $622.35   
32238   $133,124.22          $493.25   
32319   $167,927.24            $0.00   
110271  $130,795.50          $272.88   
115604  $166,542.00          $433.68   
116990  $164,740.58          $260.28   
117503  $163,778.33          $520.20   
118209  $229,221.00          $260.28   
121619  $194,541.24        $5,262.78   
121638  $119,937.08          $517.10   
121647  $136,919.76          $457.62   
121660  $151,981.92          $725.04   
121666  $201,702.96        $2,527.23   
121686  $127,500.00          $669.06   
121694  $109,579.84          $603.02   
121706  $158,799.00          $448.86   
121727  $152,466.36          $641.76   

                                            Employer  \
6223                       Landlord and Tenant Board   
9323                         Social Benefits Tribuna   
23040                              Municipal Affairs   
32225               Southlake Regional Health Centre   
32238               Southlake Regional Health Centre   
32319               Southlake Regional Health Centre   
110271  Ontario College of Art and Design University   
115604                         University of Toronto   
116990                         University of Toronto   
117503                         University of Toronto   
118209                         University of Toronto   
121619                         University of Windsor   
121638                         University of Windsor   
121647                         University of Windsor   
121660                         University of Windsor   
121666                         University of Windsor   
121686                         University of Windsor   
121694                         University of Windsor   
121706                         University of Windsor   
121727                         University of Windsor   

                                                Job Title  \
6223                             Social Benefits Tribunal   
9323     Child and Family Services Review Board and Cu...   
23040                                             Housing   
32225                                            Director   
32238                              

Unfortunately, 65 rows are affected. We need all values in the Calendar Year column to be 2016, meaning that we need to get rid of the current values somehow. It seems as though they are job titles. For example, the third row has value "Deputy Minister" in the Calendar Year column. The corresponding job title is "Housing", which surely cannot be a job title. On the other hand, we see that the row below it has job title "Director". But, the value in the Calendar Year column is "seconded to Stevenson Memorial Hospital as CEO".

From the [2015 salary disclosure page](https://www.ontario.ca/page/public-sector-salary-disclosure-2015-all-sectors-and-seconded-employees):
>Someone who is “seconded” has a job in a public sector organization (other than an Ontario government ministry), but currently works within a government ministry. The organization pays the person’s salary and benefits and the ministry reimburses the organization.

Since the dataframes at least have the same column names at this point, we will concatenate them to make our job easier for this section.

In [18]:
pss_comb = pd.concat([pss[year] for year in YEARS]).copy()

Let's take a closer look at a few people's job titles and corresponding calendar year values in attempt to find a solution to this problem.

In [19]:
pss_comb[pss_comb['Last Name'].str.contains('Fagan', case=False) & pss_comb['First Name'].str.contains('Thomas', case=False)]

Sector Last Name First Name   Salary Paid  \
5752              Ontario Public Service     FAGAN   THOMAS F   $104,654.07   
7247              Ontario Public Service     FAGAN     THOMAS   $110,957.16   
8590              Ontario Public Service     FAGAN     THOMAS   $134,556.84   
2234  Government of Ontario : Ministries     FAGAN     THOMAS  $130,123.23    
2288  Government of Ontario : Ministries     FAGAN     THOMAS   $138,193.89   
2854  Government of Ontario : Ministries     FAGAN     THOMAS  $142,724.90    
2586  Government of Ontario : Ministries     FAGAN     THOMAS   $159,197.24   
3647  Government of Ontario : Ministries     FAGAN     THOMAS   $172,187.73   
4959  Government of Ontario : Ministries     FAGAN     THOMAS   $185,823.02   
6283  Government of Ontario : Ministries     FAGAN     THOMAS   $197,394.14   
6589  Government of Ontario : Ministries     FAGAN     THOMAS   $192,951.78   
7875  Government of Ontario : Ministries     FAGAN     THOMAS   $201,608.89   
8647  Government of Ontario - Ministries     FAGAN     THOMAS     201608.89   
9062  Government of Ontario - Ministries     FAGAN     THOMAS     167073.72   
6223                      Crown Agencies     Fagan     Thomas   $126,498.77   
2529                      Crown Agencies     Fagan  Thomas F.   $143,820.61   
7638                      Crown Agencies     Fagan     Thomas  $ 153,181.89   

     Taxable Benefits                                           Employer  \
5752          $285.40                                   Attorney General   
7247          $254.85                                   Attorney General   
8590          $223.52                                   Attorney General   
2234         $231.59                                    Attorney General   
2288          $240.95                                   Attorney General   
2854         $235.49                                    Attorney General   
2586          $268.40                                   Attorney General   
3647          $301.74                                   Attorney General   
4959          $332.66                                   Attorney General   
6283          $326.25                                   Attorney General   
6589          $324.89                                   Attorney General   
7875          $289.77               Attorney General / Procureur général   
8647            277.2                                   Attorney General   
9062              0.0                                  Attorney General    
6223            $0.00                          Landlord and Tenant Board   
2529            $0.00  Landlord and Tenant Board / Commission de la l...   
7638           $ 0.00                          Landlord and Tenant Board   

                                              Job Title  \
5752                         Director, Court Operations   
7247                         Director, Court Operations   
8590                         Director, Court Operations   
2234                         Director, Court Operations   
2288                         Director, Court Operations   
2854                         Director, Court Operations   
2586                         Director, Court Operations   
3647                         Director, Court Operations   
4959                         Director, Court Operations   
6283                    Director, Diversity Initiatives   
6589                    Director, Diversity Initiatives   
7875  Director Diversity Initiatives / Directeur, in...   
8647                     Director Diversity Initiatives   
9062                   Director, Diversity Initiatives    
6223                           Social Benefits Tribunal   
2529                                    Member / Membre   
7638                                             Member   

                                      Calendar Year  
5752                                           2000  
7247                                           2001  
8590             

Fagan's job title in 2016 is very likely "Member", just as in subsequent years.

In [20]:
pss_comb[pss_comb['Last Name'].str.contains('leblanc', case=False) & pss_comb['First Name'].str.contains('Laurie', case=False)]

Sector Last Name First Name   Salary Paid  \
6715               Ontario Public Service   LEBLANC     LAURIE   $100,117.22   
8760               Ontario Public Service   LEBLANC     LAURIE   $108,753.38   
10468              Ontario Public Service   LEBLANC     LAURIE   $110,957.01   
4203   Government of Ontario : Ministries   LEBLANC     LAURIE  $112,785.31    
3080   Government of Ontario : Ministries   LEBLANC     LAURIE   $124,170.53   
3676   Government of Ontario : Ministries   LEBLANC     LAURIE  $139,155.05    
3514   Government of Ontario : Ministries   LEBLANC     LAURIE   $157,447.13   
5661   Government of Ontario : Ministries   LEBLANC     LAURIE   $167,344.55   
7314   Government of Ontario : Ministries   LEBLANC     LAURIE   $176,426.00   
12476  Government of Ontario : Ministries   LEBLANC     LAURIE   $178,463.46   
13993  Government of Ontario : Ministries   LEBLANC     LAURIE   $174,888.26   
16478  Government of Ontario : Ministries   LEBLANC     LAURIE   $188,807.05   
8089   Government of Ontario - Ministries   LEBLANC     LAURIE      207050.3   
17890  Government of Ontario - Ministries   LEBLANC     LAURIE     213291.35   
11271  Government of Ontario - Ministries   Leblanc     Laurie   $221,157.59   
10551  Government of Ontario - Ministries   Leblanc     Laurie   $233,613.69   
23040  Government of Ontario - Ministries   LeBlanc     Laurie   $242,431.56   
21917  Government of Ontario - Ministries   Leblanc     Laurie   $267,051.52   
27797  Government of Ontario - Ministries   Leblanc     Laurie  $ 288,321.07   
30911  Government of Ontario - Ministries   Leblanc     Laurie   $367,706.36   

      Taxable Benefits                                           Employer  \
6715           $281.32                        Municipal Affairs & Housing   
8760           $252.16                        Municipal Affairs & Housing   
10468          $223.52                      Municipal Affairs and Housing   
4203          $207.09                         Municipal Affairs & Housing   
3080           $210.70                                     Cabinet Office   
3676          $217.29                                      Cabinet Office   
3514           $262.75                                     Cabinet Office   
5661           $280.28           Community Safety & Correctional Services   
7314           $294.60           Community Safety & Correctional Services   
12476          $285.32                  Training, Colleges & Universities   
13993          $282.36                  Training, Colleges & Universities   
16478          $249.87  Training, Colleges & Universities / Formation,...   
8089           4418.61                                 Aboriginal Affairs   
17890         10590.92                       Municipal Affairs & Housing    
11271       $11,400.52                      Municipal Affairs and Housing   
10551       $11,467.70                        Municipal Affairs & Housing   
23040       $23,569.27                                  Municipal Affairs   
21917       $13,561.61  Ministry of Municipal Affairs / Ministère des ...   
27797      $ 13,748.18                      Ministry of Municipal Affairs   
30911       $13,068.20                Training, Colleges and Universities   

                                               Job Title    Calendar Year  
6715                            Director, Communications             2000  
8760             Director,Urb Aff & Stakeholders Rel Br.             2001  
10468            Director, Urban Aff. & Stakeholder Rel.             2002  
4203                       Director, Urb Affrs & Stk Rel             2003  
3080                                 ADM, Communications             2004  
3676                                 ADM, Communications             2005  
3514           Assistant Deputy Minister, Communications             2006  
5661   Assistant Deputy Minister, Policy & Strategic ...             2007  
7314   Assistant Deputy Minister, Policy & Strategic

It is clear that in 2016, Laurie Leblanc's job title should be "Deputy Minister". So, we could replace "Housing" with "Deputy Minister". But, the problem is that this is not a general solution. For example, is Dora Cavallo-Medved's (in the wrong_year dataframe, third row from the bottom) job title "Sessional Lecturer I" or "Course Developer"?

In [21]:
pss_comb[pss_comb['Last Name'].str.contains('levac', case=False) & pss_comb['First Name'].str.contains('jody', case=False)]

Sector Last Name First Name  \
10105   Hospitals and Boards of Public Health     LEVAC       JODY   
13496   Hospitals and Boards of Public Health     LEVAC       JODY   
16432   Hospitals and Boards of Public Health     LEVAC       JODY   
68542   Hospitals and Boards of Public Health     LEVAC       JODY   
20999   Hospitals and Boards of Public Health     LEVAC       JODY   
22835   Hospitals and Boards of Public Health     LEVAC       JODY   
23872   Hospitals and Boards of Public Health     LEVAC       JODY   
100867  Hospitals and Boards of Public Health     Levac    Jody J.   
103961    Hospitals & Boards of Public Health     Levac    Jody J.   
32225   Hospitals and Boards of Public Health     Levac    Jody J.   
30594   Hospitals and Boards of Public Health     Levac    Jody J.   
42072   Hospitals and Boards of Public Health     Levac       Jody   
46331     Hospitals & Boards of Public Health     Levac       Jody   

         Salary Paid Taxable Benefits  \
10105    $103,611.97        $6,443.50   
13496    $115,024.89        $5,869.93   
16432    $116,763.88        $6,339.81   
68542    $120,334.86        $4,690.41   
20999    $122,959.69        $6,649.28   
22835      127370.77         13498.25   
23872      132595.35           501.02   
100867   $176,593.90          $632.04   
103961   $169,940.16          $632.04   
32225    $169,367.20          $622.35   
30594    $136,176.28          $408.39   
42072   $ 227,732.10       $ 2,073.48   
46331    $225,648.98        $2,073.48   

                                                 Employer  \
10105                    Southlake Regional Health Centre   
13496                    Southlake Regional Health Centre   
16432                    Southlake Regional Health Centre   
68542                    Southlake Regional Health Centre   
20999   Southlake Regional Health Centre / Centre régi...   
22835                    Southlake Regional Health Centre   
23872                    Southlake Regional Health Centre   
100867                   Southlake Regional Health Centre   
103961                   Southlake Regional Health Centre   
32225                    Southlake Regional Health Centre   
30594                    Southlake Regional Health Centre   
42072                         Stevenson Memorial Hospital   
46331                         Stevenson Memorial Hospital   

                                                Job Title  \
10105                             Director, Mental Health   
13496                             Director, Mental Health   
16432                             Director, Mental Health   
68542                                             Manager   
20999                                  Director/Directeur   
22835                                            Director   
23872                                Director / Directeur   
100867  Director and Chief Executive Officer Stevenson...   
103961  Director; seconded to Stevenson Memorial Hospi...   
32225                                            Director   
30594                                            Director   
42072           Chief Executive Officer/Directeur General   
46331                             Chief Executive Officer   

                                          Calendar Year  
10105                                              2007  
13496                                              2008  
16432                                              2009  
68542                                              2010  
20999                                              2011  
22835                                              2012  
23872                                              2013  
100867                                             2014  
103961                                             2015  
32225    seconded to Stevenson Memorial Hospital as CEO  
30594                                              2017  
42072                                              2018  
46331              

In 2011, 2013, and 2018, the job titles are of the form "English / French", unlike the other years. Even worse, in 2011, both the employer and job title are of the form "English / French" but in 2013, only the job title is. Even the name columns are inconsistent: from 2007 to 2013, the first and last names are in upper case, unlike 2014 to 2019. From 2014 to 2017, the first name column has an initial "J.", which is not consistent with the rest of the years.

This one gives us an idea. We could append the string in the Calendar Year column to the job title column, separated by a semicolon, as in 2015.

In [22]:
pss[2016].loc[pss[2016]['Calendar Year'] != '2016', 'Job Title'] = wrong_year['Job Title'].str.cat(wrong_year['Calendar Year'], sep='; ')
pss[2016].loc[pss[2016]['Calendar Year'] != '2016', 'Calendar Year'] = '2016'
pss[2016]['Calendar Year'] = pss[2016]['Calendar Year'].astype('int')
pss[2016].dtypes

Sector              object
Last Name           object
First Name          object
Salary Paid         object
Taxable Benefits    object
Employer            object
Job Title           object
Calendar Year        int64
dtype: object

Had we dropped the Calendar Year column before examining it closely, we would have lost job title information for 65 rows in the 2016 dataframe. Therefore, it is always a good idea to examine a column before dropping it when cleaning data.

Before moving on, we will take a look at the Calendar Year values in each dataframe to make sure that there are no surprises.

In [23]:
for year in YEARS:
    if pss[year]['Calendar Year'].nunique() > 1:
        print(year, pss[year]['Calendar Year'].unique(), sep='\n')

2015
[2015 2016 2017 2018]


It looks like the Calendar Year column in the 2015 dataframe has values other than 2015. Another important lesson: if we had assumed that there were no mistakes in the Calendar Year column, we could make some serious errors. For example, if we decided to concatenate the dataframes and perform a groupby operation on the Calendar Year column, then several rows would be incorrectly grouped into 2016, 2017, and 2018.

In [24]:
wrong_year_2015 = pss[2015][pss[2015]['Calendar Year'] != 2015]
wrong_year_2015

Sector  Last Name  First Name  Salary Paid  \
76671                    Universities       Webb        Paul  $105,001.75   
111692  Other Public Sector Employers        Yue       Jenny  $114,158.53   
113433  Other Public Sector Employers  Macdonald   Christine  $100,789.02   
113434  Other Public Sector Employers     Parlee     Forrest  $108,709.89   
115175  Other Public Sector Employers     Udovic     Natasha  $129,807.63   
115176  Other Public Sector Employers     Wagner  John Clyde  $168,230.79   
115177  Other Public Sector Employers  Weisbrodt        Jorn  $207,692.37   
115248  Other Public Sector Employers      Tanos     Antonia  $102,222.20   

       Taxable Benefits                                           Employer  \
76671           $622.01                          King's University College   
111692          $131.94                                   Colleges Ontario   
113433               $-                                        Mitacs Inc.   
113434               $-                                        Mitacs Inc.   
115175          $867.92   Toronto Festival of Arts, Culture and Creativity   
115176          $867.92   Toronto Festival of Arts, Culture and Creativity   
115177        $1,272.92   Toronto Festival of Arts, Culture and Creativity   
115248        $3,664.48  Toronto Organizing Committee for the 2015 Pan ...   

                                                Job Title  Calendar Year  
76671                                           Professor           2016  
111692         Senior Manager, Finance And Administration           2016  
113433  Director, Business Development & Team Lead Ont...           2016  
113434                             Director, Partnerships           2017  
115175            Senior Director, Corporate Partnerships           2016  
115176                                 Executive Producer           2017  
115177                                  Artistic Director           2018  
115248  Senior Project Management Analyst / Analyste p...           2016

Only a few of the rows are affected. Since this data was released in 2016 for the calendar year 2015, the above values don't make sense. We will correct these now.

In [25]:
pss[2015].loc[wrong_year_2015.index, 'Calendar Year'] = 2015

Now that the Calendar Year column has the same data type across all dataframes, we will move on to examine the Salary Paid and Taxable Benefits columns. We already saw that we need to remove dollar signs and commas. But, is that everything? We will check for other non-numeric characters. In the two code cells below, we search for any non-numeric characters excluding dollar signs, commas, periods, and spaces.

In [26]:
pss_comb = pd.concat([pss[year] for year in YEARS]).copy().reset_index(drop=True)

salary_nonnum = pss_comb['Salary Paid'].str.extractall('([^$\d.,\s])').drop_duplicates()
salary_nonnum

Empty DataFrame
Columns: [0]
Index: []

In [27]:
tax_ben_nonnum = pss_comb['Taxable Benefits'].str.extractall('([^$\d.,\s])').drop_duplicates()
tax_ben_nonnum

0
       match   
780155 0      -

It looks like we have dashes in the Taxable Benefits column. We will need to remove these before converting to this column to the `float` data type. Before we do so, we should take a look at how the dash appears in the data.

In [28]:
idx = tax_ben_nonnum.reset_index(level='match', drop=True).index
pss_comb.loc[idx, 'Taxable Benefits'].unique()

array(['$-'], dtype=object)

The dash only appears in the form "$-".

Once we have replaced all dollar signs and commas, we will replace all dashes with `np.nan` since these are missing values.

In [29]:
for year in YEARS:
    pss[year]['Salary Paid'] = (pss[year]['Salary Paid']
                                .replace('[$,]', '', regex=True)
                                .replace('-', np.nan)
                                .astype('float')
                               )
    pss[year]['Taxable Benefits'] = (pss[year]['Taxable Benefits']
                                     .replace('[$,]', '', regex=True)
                                     .replace('-', np.nan)
                                     .astype('float')
                                    )

Finally, we convert the columns that have data type `object` to `string`. We may do so by calling `convert_dtypes()` on each dataframe. This will convert the columns to the best possible data types. The [Pandas documentation](https://pandas.pydata.org/pandas-docs/stable/getting_started/basics.html#dtypes) explains why this is a good idea:

> Pandas has two ways to store strings.
> 1. object dtype, which can hold any Python object, including strings.
> 1. StringDtype, which is dedicated to strings.
> 
> Generally, we recommend using StringDtype.
> Finally, arbitrary objects may be stored using the object dtype, but should be avoided to the extent possible (for performance and interoperability with other libraries and methods).

In [30]:
for year in YEARS:
    pss[year] = pss[year].convert_dtypes()
    print(year, pss[year].dtypes, sep='\n', end='\n\n')

1996
Sector               string
Last Name            string
First Name           string
Salary Paid         Float64
Taxable Benefits    Float64
Employer             string
Job Title            string
Calendar Year         Int64
dtype: object

1997
Sector               string
Last Name            string
First Name           string
Salary Paid         Float64
Taxable Benefits    Float64
Employer             string
Job Title            string
Calendar Year         Int64
dtype: object

1998
Sector               string
Last Name            string
First Name           string
Salary Paid         Float64
Taxable Benefits    Float64
Employer             string
Job Title            string
Calendar Year         Int64
dtype: object

1999
Sector               string
Last Name            string
First Name           string
Salary Paid         Float64
Taxable Benefits    Float64
Employer             string
Job Title            string
Calendar Year         Int64
dtype: object

2000
Sector             

Each column now has the desired data type. Since we are writing this data to a CSV, we will need to convert the data types again when we explore this data in a different notebook. So, why convert data types here? We ran into several issues while attempting to do so. The idea is to take care of those issues here instead of in an analysis notebook.

<a id="missing-values"></a>
[Return to table of contents](#table-of-contents)

# Missing values

In [31]:
for year in YEARS:
    if pss[year].isna().sum().sum() != 0:
        print(year, pss[year].isna().sum(), sep='\n', end='\n\n')

1996
Sector              0
Last Name           0
First Name          1
Salary Paid         0
Taxable Benefits    0
Employer            0
Job Title           0
Calendar Year       0
dtype: int64

1997
Sector              0
Last Name           0
First Name          0
Salary Paid         0
Taxable Benefits    0
Employer            0
Job Title           1
Calendar Year       0
dtype: int64

1998
Sector              0
Last Name           0
First Name          1
Salary Paid         0
Taxable Benefits    0
Employer            0
Job Title           0
Calendar Year       0
dtype: int64

2013
Sector              0
Last Name           0
First Name          1
Salary Paid         0
Taxable Benefits    0
Employer            0
Job Title           0
Calendar Year       0
dtype: int64

2015
Sector                 0
Last Name              0
First Name             0
Salary Paid            0
Taxable Benefits    6980
Employer               0
Job Title              0
Calendar Year          0
dtype: int64

2

Apart from the 6980 in the 2015 dataframe, there are not many missing values. Let's first find out why there are so many missing values in the Taxable Benefits column in 2015. Maybe those missing values are supposed to be zeros.

In [32]:
pss[2015]['Taxable Benefits'].eq(0).sum()

0

It seems as if every employee received some amount of taxable benefits in 2015. This seems unlikely, but we could take a look at the data from other years and compare.

In [33]:
pss_comb = (pd.concat([pss[year] for year in YEARS])
            .copy()
            .reset_index(drop=True)
           )
no_tax_ben = (pss_comb
              .loc[pss_comb['Taxable Benefits'].eq(0), 'Calendar Year']
              .value_counts()
              .reindex(list(YEARS), fill_value=0)
              .to_frame()
              .reset_index()
              .rename(columns={'index': 'Calendar Year', 'Calendar Year': 'Number of employees'})
             )

In [34]:
fig = px.scatter(no_tax_ben, x='Calendar Year', y='Number of employees')
fig.update_traces(mode='lines+markers',
                  hovertemplate=
                  '<b>%{x}</b><br>'+
                  'Number of employees: <b>%{y}</b>'
                 )
fig.update_layout(title='Number of employees that did not receive taxable benefits by calendar year',
                  xaxis_title='Calendar Year',
                  yaxis_title="Number of employees",
                  yaxis_tickformat=',',
                  hoverlabel_bgcolor="white",
                  hoverlabel_font_size=14,
                  hovermode="x",
                  yaxis_zerolinecolor='grey',
                  yaxis_zerolinewidth=1
                 )
fig.show(config=CONFIG)

Based on the above plot, we can reasonably assume that the missing values for the 2015 Taxable Benefits column must be zeros. We will fill in the missing values and then plot the data again.

In [35]:
pss[2015]['Taxable Benefits'].fillna(0.0, inplace=True)
no_tax_ben.loc[no_tax_ben['Calendar Year'].eq(2015), 'Number of employees'] = pss[2015]['Taxable Benefits'].eq(0).sum()

In [36]:
fig = px.scatter(no_tax_ben, x='Calendar Year', y='Number of employees')
fig.update_traces(mode='lines+markers',
                  hovertemplate=
                  '<b>%{x}</b><br>'+
                  'Number of employees: <b>%{y}</b>'
                 )
fig.update_layout(title='Number of employees that did not receive taxable benefits by calendar year',
                  xaxis_title='Calendar Year',
                  yaxis_title="Number of employees",
                  yaxis_tickformat=',',
                  hoverlabel_bgcolor="white",
                  hoverlabel_font_size=14,
                  hovermode="x",
                  yaxis_zerolinecolor='grey',
                  yaxis_zerolinewidth=1
                 )
fig.show(config=CONFIG)

The plot makes more sense now. The number of employees that did not receive taxable benefits in 2015 is between the corresponding values for 2014 and 2016.

The 2016 data has one missing value in the Taxable Benefits column.

In [37]:
null_2016 = pss[2016][pss[2016]['Taxable Benefits'].isna()]
null_2016

Sector  Last Name First Name  Salary Paid  \
84270  Other Public Sector Employers  Malenfant      James    152545.47   

       Taxable Benefits                                 Employer  \
84270              <NA>  Independent Electricity System Operator   

                                               Job Title  Calendar Year  
84270  Senior Analyst - Compliance and Dispute Resolu...           2016

We will look for "James Malenfant" in other years in hopes of being able to impute this value.

In [38]:
pss_comb[pss_comb['Last Name'].eq('Malenfant') & pss_comb['First Name'].eq('James')]

Sector  Last Name First Name  Salary Paid  \
892807   Other Public Sector Employers  Malenfant      James    107325.45   
980270   Other Public Sector Employers  Malenfant      James    152545.47   
1114900  Other Public Sector Employers  Malenfant      James    153909.53   
1256693  Other Public Sector Employers  Malenfant      James    155282.66   
1417573  Other Public Sector Employers  Malenfant      James    158596.56   

         Taxable Benefits                                           Employer  \
892807               <NA>            Independent Electricity System Operator   
980270               <NA>            Independent Electricity System Operator   
1114900               0.0  Independent Electricity System Operator / Opér...   
1256693               0.0            Independent Electricity System Operator   
1417573               0.0            Independent Electricity System Operator   

                                                 Job Title  Calendar Year  
892807   Senior Analyst - Compliance & Dispute Resoluti...           2015  
980270   Senior Analyst - Compliance and Dispute Resolu...           2016  
1114900  Senior Analyst, Compliance and Dispute Resolut...           2017  
1256693  Senior Analyst - Compliance and Dispute Resolu...           2018  
1417573  Senior Analyst - Compliance and Dispute Resolu...           2019

Since James Malenfant does not usually receive taxable benefits, it is reasonable to assume that he didn't receive taxable benefits in 2016 either.

In [39]:
pss[2016].loc[null_2016.index, 'Taxable Benefits'] = 0.0

The 2013 data has one missing value in the First Name column.

In [40]:
null_2013 = pss[2013][pss[2013]['First Name'].isna()]
null_2013

Sector Last Name First Name  Salary Paid  Taxable Benefits  \
4615  Crown Agencies        LI       <NA>    108421.82             47.18   

              Employer                        Job Title  Calendar Year  
4615  eHealth Ontario   Privacy and Security Architect            2013

In [41]:
pss_comb[pss_comb['Last Name'].str.contains('^li$', case=False) & pss_comb['Employer'].str.contains('eHealth') & pss_comb['Job Title'].str.contains('privacy', case=False)]

Sector Last Name First Name  Salary Paid  Taxable Benefits  \
575124  Crown Agencies        LI       <NA>    108421.82             47.18   
683346  Crown Agencies        Li         Na    106867.01             42.12   

                Employer                                          Job Title  \
575124  eHealth Ontario                     Privacy and Security Architect    
683346   eHealth Ontario  Privacy and Security Architect / Architecte de...   

        Calendar Year  
575124           2013  
683346           2014

We can safely assume that Li's first name is Na based on the record from 2014. It looks like the first name "NA" was marked as a missing value in the 2013 data. The first name "Na" in the 2014 data was not marked as a missing value. From the `pandas.read_csv` [documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html):

> **na_values: scalar, str, list-like, or dict, optional**   
> Additional strings to recognize as NA/NaN. If dict passed, specific per-column NA values. By default the following values are interpreted as NaN: ‘’, ‘#N/A’, ‘#N/A N/A’, ‘#NA’, ‘-1.#IND’, ‘-1.#QNAN’, ‘-NaN’, ‘-nan’, ‘1.#IND’, ‘1.#QNAN’, ‘<NA>’, ‘N/A’, ‘NA’, ‘NULL’, ‘NaN’, ‘n/a’, ‘nan’, ‘null’.
    
This means that LI's first name "NA" in the 2013 data was likely never really a missing value to begin with. Let's find out what the original first name value was in 2013. We can do so by passing the argument `False` to the `na_filter` parameter when calling `pd.read_csv`.

In [42]:
pss2013 = pd.read_csv(path/'pssd-en-2013.csv', na_filter=False)
pss2013[pss2013['Last Name'].str.contains('^li$', case=False) & pss2013['Employer'].str.contains('eHealth') & pss2013['Job Title'].str.contains('privacy', case=False)]

Sector Last Name First Name  Salary Paid  Taxable Benefits  \
4615  Crown Agencies        LI         NA    108421.82             47.18   

              Employer                        Job Title  Calendar Year  
4615  eHealth Ontario   Privacy and Security Architect            2013

Indeed we see that the first name was originally "NA", but was marked as a missing value. We will set the missing first name to "NA" instead of "Na" in order to be consistent, since names in the 2013 data seem to be in upper case only.

In [43]:
pss[2013].loc[null_2013.index, 'First Name'] = 'NA'

The 1998 data has one missing value in the first name column.

In [44]:
null_1998 = pss[1998][pss[1998]['First Name'].isna()]
null_1998

Sector Last Name First Name  Salary Paid  Taxable Benefits  \
595  Crown Agencies  DONNELLY       <NA>     100813.0           21526.0   

          Employer                       Job Title  Calendar Year  
595  Ontario Hydro  Mgr, Labour Reltns, IMO/CMO/N-           1998

In [45]:
pss_comb[pss_comb['Last Name'].str.contains('donnelly', case=False) & pss_comb['Employer'].str.contains('hydro', case=False)]

Sector Last Name First Name  Salary Paid  \
10474                        Crown Agencies  DONNELLY       <NA>     100813.0   
17761  Hydro One & Ontario Power Generation  DONNELLY         N.    113935.62   

       Taxable Benefits       Employer                       Job Title  \
10474           21526.0  Ontario Hydro  Mgr, Labour Reltns, IMO/CMO/N-   
17761           1098.59      Hydro One        Labour Relations Manager   

       Calendar Year  
10474           1998  
17761           1999

Based on the record from 1999, we can safely assume that Donnelly's first initial is "N". However, just as we saw above, it may be the case that `pd.read_csv` marked Donnelly's first name as a missing value.

In [46]:
pss1998 = pd.read_csv(path/'en-1998-pssd.csv', encoding='latin1', na_filter=False)
pss1998[pss1998['Last Name'].str.contains('donnelly', case=False) & pss1998['Employer'].str.contains('hydro', case=False)]

Sector Last Name First Name  Salary Paid Taxable Benefits  \
595  Crown Agencies  DONNELLY         NA  $100,813.00       $21,526.00   

          Employer                       Job Title  Calendar Year  
595  Ontario Hydro  Mgr, Labour Reltns, IMO/CMO/N-           1998

Just like before, a first name of "NA" was marked as a missing value. Donnelly's first name is probably not "NA" though. It's more likely that these are initials. It's not uncommon in the 1998 data to have 2 letter initials in the First Name column.

In [47]:
pss[1998][pss[1998]['First Name'].str.len().eq(2)].head()

Sector Last Name First Name  Salary Paid  Taxable Benefits  \
266  Crown Agencies  ACCHIONE         PN     107527.0           10377.0   
269  Crown Agencies     ADAMS         RJ     107327.0           28414.0   
273  Crown Agencies    AITKEN         RB     109803.0             463.0   
281  Crown Agencies     ALLEN         MI     100884.0             266.0   
285  Crown Agencies  ANDERSON         WA     100561.0             330.0   

          Employer                                Job Title  Calendar Year  
266  Ontario Hydro                     Manager, Programming           1998  
269  Ontario Hydro           Manager, Operations Production           1998  
273  Ontario Hydro  Shift Maintenance Supervisor-Mechanical           1998  
281  Ontario Hydro     Intermediate Mail and Services Clerk           1998  
285  Ontario Hydro  Inspection and Maintenance Technician I           1998

Therefore, we'll set the missing first name to "NA", as it was originally.

In [48]:
pss[1998].loc[null_1998.index, 'First Name'] = 'NA'

The 1997 data has one missing value in the Job Title column.

In [49]:
null_1997 = pss[1997][pss[1997]['Job Title'].isna()]
null_1997

Sector Last Name First Name  Salary Paid  Taxable Benefits  \
963  Crown Agencies    Walker        G J    104932.87            403.32   

          Employer Job Title  Calendar Year  
963  Ontario Hydro      <NA>           1997

In [50]:
pss_comb[pss_comb['Last Name'].str.contains('walker', case=False) & pss_comb['Employer'].str.contains('ontario hydro', case=False)]

Sector Last Name First Name  Salary Paid  Taxable Benefits  \
2302   CROWN AGENCIES    WALKER          J     103850.0             416.0   
2303   CROWN AGENCIES    WALKER          J     112519.0             536.0   
2304   CROWN AGENCIES    WALKER          W     136763.0             712.0   
5462   Crown Agencies    Walker  Colleen M    160369.87           1267.71   
5463   Crown Agencies    Walker        D G    124642.07            418.83   
5464   Crown Agencies    Walker        G J    104932.87            403.32   
5465   Crown Agencies    Walker        J H     113798.7            487.05   
5466   Crown Agencies    Walker    Wayne A     123669.6           2011.86   
11457  Crown Agencies    WALKER          G     113573.0             489.0   
11458  Crown Agencies    WALKER          G     126936.0             480.0   
11459  Crown Agencies    WALKER          J     117236.0             555.0   
11460  Crown Agencies    WALKER         WA     147962.0           66616.0   

            Employer                                 Job Title  Calendar Year  
2302   Ontario Hydro               AUTHORIZED NUCLEAR OPERATOR           1996  
2303   Ontario Hydro   SHIFT OPERATING SUPERVISOR (AUTHORIZED)           1996  
2304   Ontario Hydro  SERVICE MGR, INFORMATION TECHNOLOGY SRVC           1996  
5462   Ontario Hydro                       EXECUTIVE ASSISTANT           1997  
5463   Ontario Hydro    SENIOR ENGINEER/SCIENTIST/TECH OFFICER           1997  
5464   Ontario Hydro                                      <NA>           1997  
5465   Ontario Hydro   SHIFT OPERATING SUPERVISOR (AUTHORIZED)           1997  
5466   Ontario Hydro   SERVICE MGR, INFORMATION TECHNOLOGY SRV           1997  
11457  Ontario Hydro                Maintenance Superintendent           1998  
11458  Ontario Hydro    Senior Engineer/Scientist/Tech Officer           1998  
11459  Ontario Hydro   Shift Operating Supervisor (Authorized)           1998  
11460  Ontario Hydro        Service Mgr, Info Technology Serv.           1998

Since they have the same job title, "Walker, D G" from 1997 and "Walker, G" from 1998 seem to be the same person. The other "Walker, G" has the job title "Maintenance Superintendent" in 1998, which is one year after "Walker, G J" in 1997. My best guess then, is to say that "Walker, G J" has job title "Maintenance Superintendent". This [external page](https://www.ontariosunshinelist.com/people/fqpyng) seems to support my claim.

In [51]:
pss[1997].loc[null_1997.index, 'Job Title'] = 'Maintenance Superintendent'

The 1996 data has one missing value in the First Name column.

In [52]:
null_1996 = pss[1996][pss[1996]['First Name'].isna()]
null_1996

Sector Last Name First Name  Salary Paid  Taxable Benefits  \
1714  MUNICIPALITIES  Yearwood       <NA>     116371.0             978.0   

                                  Employer  \
1714  Municipality of Metropolitan Toronto   

                               Job Title  Calendar Year  
1714  Deputy Commissioner of Corp. Serv.           1996

In [53]:
pss_comb[pss_comb['Last Name'].str.contains('yearwood', case=False) & pss_comb['Calendar Year'].le(2010)]

Sector Last Name First Name  Salary Paid  \
1714                 MUNICIPALITIES  Yearwood       <NA>     116371.0   
130927                 Universities  YEARWOOD     LOUISE    109492.62   
355991  Municipalities and Services  YEARWOOD   MICHELLE    102838.21   

        Taxable Benefits                              Employer  \
1714               978.0  Municipality of Metropolitan Toronto   
130927            390.14                    Ryerson University   
355991            405.75                      City of Brampton   

                                     Job Title  Calendar Year  
1714        Deputy Commissioner of Corp. Serv.           1996  
130927              Director, Development, C&D           2005  
355991  Senior Advisor, Performance Management           2010

This is not helpful. Let's find out what the original value of the first name was.

In [54]:
pss1996 = pd.read_csv(path/'en-1996-pssd.csv', encoding='latin1', na_filter=False)
pss1996[pss1996['Last Name'].str.contains('yearwood', case=False)]

Sector Last Name First Name  Salary Paid Taxable Benefits  \
1714  MUNICIPALITIES  Yearwood         NA  $116,371.00          $978.00   

                                  Employer  \
1714  Municipality of Metropolitan Toronto   

                               Job Title  Calendar Year Unnamed: 8  
1714  Deputy Commissioner of Corp. Serv.           1996

Again, a first name of "NA" was interpreted to be a missing value by the `pd.read_csv` function. This is the last missing value. If we had many missing values, we would have to deal with them in a less cumbersome way. For example, we could specify that we only want empty strings to be interpreted as missing values when reading in the data with `pd.read_csv`.

Just like before, we'll keep the original first name of "NA", since it's not uncommon to have 2 letter initials in the 1996 data.

In [55]:
pss[1996][pss[1996]['First Name'].str.len().eq(2)].head()

Sector  Last Name First Name  Salary Paid  \
605   ONTARIO PUBLIC SERVICE    WEBSTER         AR     123825.0   
820           MUNICIPALITIES     Baxter         SK     106526.0   
851           MUNICIPALITIES   Wichmann         WM     105603.0   
892           MUNICIPALITIES       Lobo         J.     116588.0   
1072               HOSPITALS  Arsenault         D.     146827.0   

      Taxable Benefits                              Employer  \
605                0.0                      Attorney General   
820              374.0                       City of Toronto   
851               33.0                       City of Toronto   
892              691.0   Corporation of the City of Hamilton   
1072            8077.0  Hamilton Health Sciences Corporation   

                                         Job Title  Calendar Year  
605                          Provincial Judge, OCJ           1996  
820                                     City Clerk           1996  
851                Director, Engineering & Surveys           1996  
892   Commissioner of Public Works & Traffic Serv.           1996  
1072                                Vice-President           1996

In [56]:
pss[1996].loc[null_1996.index, 'First Name'] = 'NA'

<a id="concat-data"></a>
[Return to table of contents](#table-of-contents)

# Concatenating the dataframes

In [57]:
pss_comb = pd.concat([pss[year] for year in YEARS]).copy()

The data spans from 1996 to 2019 and as we saw throughout this notebook, there were many inconsistencies. Here is just one of many examples:

In [58]:
pss_comb[pss_comb['Last Name'].str.contains('malenfant', case=False) & pss_comb['First Name'].str.contains('andrew', case=False)]

Sector  Last Name    First Name  Salary Paid  \
33002  Municipalities and Services  MALENFANT  ANDREW DEREK     110851.7   
33427  Municipalities and Services  MALENFANT  ANDREW DEREK    115988.74   
48487  Municipalities and Services  MALENFANT  ANDREW DEREK    116828.38   
54019  Municipalities and Services  MALENFANT  ANDREW DEREK    108888.82   
46296  Municipalities and Services  Malenfant  Andrew Derek    117105.33   
43728    Municipalities & Services  Malenfant  Andrew Derek    137741.05   
56955  Municipalities and Services  Malenfant  Andrew Derek    131296.57   
69543  Municipalities and Services  Malenfant  Andrew Derek    131755.36   
72618  Municipalities and Services  Malenfant  Andrew Derek     135177.3   
80388    Municipalities & Services  Malenfant        Andrew    146555.36   

       Taxable Benefits                          Employer  \
33002            324.09  City of Toronto - Police Service   
33427            751.24  City of Toronto - Police Service   
48487            686.42  City of Toronto - Police Service   
54019            759.93  City of Toronto - Police Service   
46296            754.78  City of Toronto - Police Service   
43728            818.86  City of Toronto - Police Service   
56955            981.89  City of Toronto - Police Service   
69543           1000.74  City of Toronto - Police Service   
72618           1008.17  City of Toronto - Police Service   
80388           1023.96  City Of Toronto - Police Service   

                           Job Title  Calendar Year  
33002               Police Constable           2010  
33427  Plainclothes Police Constable           2011  
48487  Plainclothes Police Constable           2012  
54019               Police Constable           2013  
46296               Police Constable           2014  
43728               Police Constable           2015  
56955               Police Constable           2016  
69543               Police Constable           2017  
72618               Police Constable           2018  
80388               Police Constable           2019

I think this is a good example because it displays several inconsistencies. It is clear that these records all belong to the same person. In the Sector column, some rows use the ampersand (&) instead of the word "and". In the Last Name and First Name columns, some values are in upper case while others are not. Also, in 2019, the first name is just "Andrew" instead of "Andrew Derek".

It would be nice if each person were given some sort of unique id. This way, even is a person's name changes, or if we see something like the example above, we would be able to easily tell whether two people on the list are the same person.

It is much easier to work with one dataframe as opposed to 24 and the calendar year column can always be used to separate the data again if needed. Another reason one CSV file is preferred over 24 is that file descriptions and column descriptions in the resulting dataset will only need to be specified one time instead of 24 times. I will also add a note in the dataset description about the inconsistency of the data across calendar years.

<a id="write-to-csv"></a>
[Return to table of contents](#table-of-contents)

# Write to CSV

In [59]:
pss_comb

Sector   Last Name First Name  Salary Paid  \
0       OTHER PUBLIC SECTOR EMPLOYERS     Kendall      Perry     194890.4   
1       OTHER PUBLIC SECTOR EMPLOYERS        Rehm    Juergen    115603.62   
2       OTHER PUBLIC SECTOR EMPLOYERS        Room      Robin    149434.48   
3              ONTARIO PUBLIC SERVICE        KNOX      KEN W    109382.92   
4                           HOSPITALS       Cliff      Bruce     110309.0   
...                               ...         ...        ...          ...   
166972                   Universities       Zoidl    Georg R    226479.06   
166973                   Universities        Zryd    Michael     171004.6   
166974                   Universities   Zumbansen     Peer C    193639.44   
166975                   Universities       Zwick     Detlev    256491.58   
166976                   Universities  Zylberberg       Joel     146093.4   

        Taxable Benefits                             Employer  \
0                 711.24        Addiction Research Foundation   
1                 403.41        Addiction Research Foundation   
2                 512.58        Addiction Research Foundation   
3                4921.68   Agriculture,Food and Rural Affairs   
4                 3157.0  Ajax and Pickering General Hospital   
...                  ...                                  ...   
166972           1104.16                      York University   
166973            910.84                      York University   
166974           1104.16                      York University   
166975           1104.16                      York University   
166976            701.64                      York University   

                                                Job Title  Calendar Year  
0                                         President & CEO           1996  
1       Dir., Soc. Eval. Research & Act. Dir., Clin. R...           1996  
2          V.P., Research & Coordinator, Intern. Programs           1996  
3                                         Deputy Minister           1996  
4                                         President & CEO           1996  
...                                                   ...            ...  
166972                                          Professor           2019  
166973                                Associate Professor           2019  
166974                                          Professor           2019  
166975               Associate Dean / Associate Professor           2019  
166976        Assistant Professor / Canada Research Chair           2019  

[1470706 rows x 8 columns]

We sort the data as a final step before writing the data to a CSV file.

In [60]:
pss_comb.sort_values(['Calendar Year', 'Sector', 'Employer', 'Last Name', 'First Name']).to_csv('pssd.csv', index=False)